genie-agent/src/main/java/com/netflix/genie/agent/execution/statemachine/listeners/JobExecutionListenerAdapter.java

```diff
@@ -0,0 +1,120 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.statemachine.listeners;
+
+import com.netflix.genie.agent.execution.statemachine.Events;
+import com.netflix.genie.agent.execution.statemachine.States;
+import org.springframework.messaging.Message;
+import org.springframework.statemachine.StateContext;
+import org.springframework.statemachine.StateMachine;
+import org.springframework.statemachine.action.Action;
+import org.springframework.statemachine.state.State;
+import org.springframework.statemachine.transition.Transition;
+
+/**
+ * Adapter for implementations of JobExecutionListener.
+ * Avoid the need to override all methods with empty implementations.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+class JobExecutionListenerAdapter implements JobExecutionListener {
```

Comment:
I suppose that could work, I went instead with the pattern they seem to be using in Spring.
I think an advantage of this is: If you add a method to the interface, the classes implementing it *must* be updated (to implement that method), whereas the classes subclassing the adapter don't.
If we go with `default` in the interface, all implementors will keep working and ignore the new method.
At the moment we have one of each (LoggingListener that should implement all methods, and JobStatus listener which only overrides one)

Reply:
The flip side of this is that any change to the interface requires changes to all implementations (including ones you don't know about if they don't extend your base class) and breaks builds. In this case why not go with an Abstract class instead of interface entirely? I assume this is cause you want to keep it package protected.